File: typescript-examples/setup-hooks/api/demo-hook.ts
import type { Page, BrowserContext } from "playwright";
import { attemptStore } from "@intuned/runtime";
import { goToUrl } from "@intuned/browser";
import {
  demoHookParamsSchema,
  type DemoHookParams,
  type HookDemoResult,
} from "../utils/typesAndSchemas";

export default async function handler(
  params: DemoHookParams,
  page: Page,
  _context: BrowserContext
): Promise<HookDemoResult> {
  // Validate parameters using Zod schema
  const validationResult = demoHookParamsSchema.safeParse(params);

  if (!validationResult.success) {
    const errors = validationResult.error.errors
      .map((err) => `${err.path.join(".")}: ${err.message}`)
      .join(", ");
    throw new Error(`Parameter validation failed: ${errors}`);
  }

  const { message = "Hello from Setup Hooks!" } = validationResult.data;
  console.log("=== Setup Hooks Demo API ===\n");

  // Step 1: Retrieve data that was stored by the setup hook
  console.log("📦 Retrieving data from setup hook...\n");

  const cdpUrl = attemptStore.get("cdpUrl") as string;
  const apiName = attemptStore.get("apiName") as string;
  const apiParameters = attemptStore.get("apiParameters");
  const executionStartTime = attemptStore.get("executionStartTime") as string;
  const config = attemptStore.get("config");
  const userAgent = attemptStore.get("userAgent") as string;

  console.log("✓ Data retrieved from attemptStore:");
  console.log(`  - CDP URL: ${cdpUrl}`);
  console.log(`  - API Name: ${apiName}`);
  console.log(`  - Execution Start: ${executionStartTime}`);
  console.log(`  - User Agent: ${userAgent}`);
  console.log(`  - Config:`, config);

  // Step 2: Use the page (connected via CDP thanks to the hook)
  console.log("\n🌐 Navigating to example.com...");
  await goToUrl({
    page,
    url: "https://example.com",
  })
  const pageTitle = await page.title();
  console.log(`✓ Page Title: ${pageTitle}`);

  // Step 3: Calculate execution time
  const endTime = new Date();
  const startTime = new Date(executionStartTime);
  const executionMs = endTime.getTime() - startTime.getTime();
  const executionTime = `${executionMs}ms`;
  console.log(`\n⏱️  Total Execution Time: ${executionTime}`);

  console.log("\n✅ Setup Hooks Demo Complete!\n");

  return {
    message,
    dataFromHook: {
      cdpUrl,
      apiName,
      apiParameters,
      executionStartTime,
      config,
      userAgent,
    },
    executionTime,
    pageTitle,
  };
}

File: python-examples/setup-hooks/api/demo-hook.py
from datetime import datetime

from intuned_browser import go_to_url
from intuned_runtime import attempt_store
from playwright.async_api import Page
from pydantic import ValidationError
from utils.types_and_schemas import (
    DataFromHook,
    DemoHookParams,
    HookDemoResult,
)


async def automation(page: Page, params: dict | None = None, **_kwargs) -> dict:
    """
    Setup Hooks Demo API

    This function demonstrates how to:
    1. Retrieve data stored by the setup hook
    2. Use the stored CDP URL for browser operations
    3. Calculate execution time using hook timestamps
    """
    print("=== Setup Hooks Demo API ===\n")

    # Validate parameters using Pydantic schema
    if not params:
        params = {}

    try:
        validated_params = DemoHookParams(**params)
    except ValidationError as e:
        errors = ", ".join([f"{err['loc'][0]}: {err['msg']}" for err in e.errors()])
        raise ValueError(f"Parameter validation failed: {errors}")

    message = validated_params.message or "Hello from Setup Hooks!"

    # Step 1: Retrieve data that was stored by the setup hook
    print("📦 Retrieving data from setup hook...\n")

    cdp_url = attempt_store.get("cdp_url")
    api_name = attempt_store.get("api_name")
    api_parameters = attempt_store.get("api_parameters")
    execution_start_time = attempt_store.get("execution_start_time")
    config = attempt_store.get("config")
    user_agent = attempt_store.get("user_agent")

    print("✓ Data retrieved from attempt_store:")
    print(f"  - CDP URL: {cdp_url}")
    print(f"  - API Name: {api_name}")
    print(f"  - Execution Start: {execution_start_time}")
    print(f"  - User Agent: {user_agent}")
    print(f"  - Config: {config}")

    # Step 2: Use the page (connected via CDP thanks to the hook)
    print("\n🌐 Navigating to example.com...")
    await go_to_url(page, url="https://example.com")
    page_title = await page.title()
    print(f"✓ Page Title: {page_title}")

    # Step 3: Calculate execution time
    end_time = datetime.now()
    start_time = datetime.fromisoformat(execution_start_time)
    execution_ms = int((end_time - start_time).total_seconds() * 1000)
    execution_time = f"{execution_ms}ms"
    print(f"\n⏱️  Total Execution Time: {execution_time}")

    print("\n✅ Setup Hooks Demo Complete!\n")

    result = HookDemoResult(
        message=message,
        data_from_hook=DataFromHook(
            cdp_url=cdp_url,
            api_name=api_name,
            api_parameters=api_parameters,
            execution_start_time=execution_start_time,
            config=config,
            user_agent=user_agent,
        ),
        execution_time=execution_time,
        page_title=page_title,
    )

    # Use mode='json' to serialize all Pydantic types to JSON-compatible primitives
    return result.model_dump(mode="json")
